core/src/nodes/handled_node.rs

```diff
@@ -362,4 +425,227 @@ mod tests {
+        });
+    }
+
+    #[test]
+    fn handler_emits_outbound_closed_when_opening_new_substream_on_closed_node() {
+        let open_event = Some(NodeHandlerEvent::OutboundSubstreamRequest(456));
+        let mut handled = TestBuilder::new()
+            .with_muxer_inbound_state(DummyConnectionState::Pending)
+            .with_muxer_outbound_state(DummyConnectionState::Closed)
+            .with_handler_state(HandlerState::Ready(open_event))
+            .handled_node();
+
+        set_next_handler_outbound_state(
+            &mut handled,
+            HandlerState::Ready(Some(NodeHandlerEvent::Custom(Event::Custom("pear"))))
+        );
+        handled.poll().expect("poll works");
+        assert_eq!(handled.handler.events, vec![Event::OutboundClosed]);
+    }
+
+    #[test]
+    fn poll_returns_not_ready_when_node_stream_and_handler_is_not_ready() {
+        let mut handled = TestBuilder::new()
+            .with_muxer_inbound_state(DummyConnectionState::Closed)
+            .with_muxer_outbound_state(DummyConnectionState::Closed)
+            .with_open_substream(12)
+            .with_handler_state(HandlerState::NotReady)
+            .handled_node();
+
+        // Under the hood, this is what happens when calling `poll()`:
+        // - we reach `node.poll_inbound()` and because the connection is
+        //   closed, `inbound_finished` is set to true.
+        // - an Async::Ready(NodeEvent::InboundClosed) is yielded (also calls
+        //   `inject_inbound_close`, but that's irrelevant here)
+        // - back in `poll()` we cal `handler.poll()` which does nothing because
+        //   `HandlerState` is `NotReady`: loop continues
+        // - polls the node again which now skips the inbound block because
+        //   `inbound_finished` is true.
+        // - Now `poll_outbound()` is called which returns `Async::Ready(None)`
+        //   and sets `outbound_finished` to true. …calls destroy_outbound and
+        //   yields Ready(OutboundClosed) …so the HandledNode calls
+        //   `inject_outbound_closed`.
+        // - Now we have `inbound_finished` and `outbound_finished` set (and no
+        //   more outbound substreams).
+        // - Next we poll the handler again which again does nothing because
+        //   HandlerState is NotReady (and the node is still there)
+        // - HandledNode polls the node again: we skip inbound and there are no
+        //   more outbound substreams so we skip that too; the addr is now
+        //   Resolved so that part is skipped too
+        // - We reach the last section and the NodeStream yields Async::Ready(None)
+        // - Back in HandledNode the Async::Ready(None) triggers a shutdown
+        // – …and causes the Handler to yield Async::Ready(None)
+        // – which in turn makes the HandledNode to yield Async::Ready(None) as well
+        assert_matches!(handled.poll(), Ok(Async::Ready(None)));
+        assert_eq!(handled.handler.events, vec![
```

Comment:
I don't think we should access private fields. The tests should only use the public API.
However I think it would be fine to add `handler(&self)` and `handler_mut(&mut self)` methods to `HandledNode`.